File: app/src/main/java/org/deiverbum/app/viewmodel/FileViewModel.java
package org.deiverbum.app.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import org.deiverbum.app.data.wrappers.CustomException;
import org.deiverbum.app.data.wrappers.DataWrapper;
import org.deiverbum.app.model.Book;
import org.deiverbum.app.repository.FileRepository;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class FileViewModel extends ViewModel {
    final FileRepository mRepository;

    @Inject
    public FileViewModel(FileRepository repository) {
        mRepository = repository;
    }

    public LiveData<DataWrapper<Book, CustomException>> getBook(String rawPath) {
        return mRepository.getBook(rawPath);
    }


}

File: app/src/main/java/org/deiverbum/app/presentation/file/FileViewModel.kt
package org.deiverbum.app.presentation.file
import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.deiverbum.app.data.networking.CoroutineDispatcherProvider
import org.deiverbum.app.domain.model.FileRequest
import org.deiverbum.app.domain.usecase.GetFileUseCase
import org.deiverbum.app.util.ExceptionParser
import javax.inject.Inject

/**
 * <p>ViewModel para el contenido proveniente de archivos locales.</p>
 *
 * @author A. Cedano
 * @since 2023.3
 */
@HiltViewModel
class FileViewModel @Inject constructor(
    private val getFileUseCase: GetFileUseCase,
    private val coroutineDispatcherProvider: CoroutineDispatcherProvider
) : ViewModel() {

    private val _uiState = MutableStateFlow<FileUiState>(FileUiState.Empty)
    val uiState: StateFlow<FileUiState> = _uiState

    fun loadData(fileRequest: FileRequest) {
        _uiState.value = FileUiState.Loading
        viewModelScope.launch(coroutineDispatcherProvider.IO()) {
            try {
                //val requestParam = FileRequest(fileRequest)
                val result = getFileUseCase.execute(fileRequest)
                _uiState.value = FileUiState.Loaded(FileItemUiState(result))
            } catch (error: Exception) {
                _uiState.value = FileUiState.Error(ExceptionParser.getMessage(error))
            }
        }
    }

    sealed class FileUiState {
        object Empty : FileUiState()
        object Loading : FileUiState()
        class Loaded(val itemState: FileItemUiState) : FileUiState()
        class Error(@StringRes val message: Int) : FileUiState()
    }
}